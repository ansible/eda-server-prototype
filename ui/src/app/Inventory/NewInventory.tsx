import { Card, CardBody as PFCardBody, Form, PageSection, Title, ValidatedOptions } from '@patternfly/react-core';
import { useHistory } from 'react-router-dom';
import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import { TopToolbar } from '@app/shared/top-toolbar';
import { useIntl } from 'react-intl';
import sharedMessages from '../messages/shared.messages';
import 'ace-builds/src-noconflict/theme-kuroir';
import { addNotification } from '@redhat-cloud-services/frontend-components-notifications';
import { useDispatch } from 'react-redux';
import { addInventory } from '@app/API/Inventory';
import InventoryForm from './shared/InventoryForm';

const CardBody = styled(PFCardBody)`
  white-space: pre-wrap;
`;
const NewInventory: React.FunctionComponent = () => {
  const history = useHistory();
  const intl = useIntl();
  const dispatch = useDispatch();
  const [error, setError] = useState(null);

  const [isSubmitting, setIsSubmitting] = useState<boolean>(false);
  const validateName = (value: string) => {
    if (!value || value.length < 1) {
      throw new Error('Name is a required field');
    }
  };

  const validateInventory = (value: string) => {
    if (!value || value.length < 1) {
      throw new Error('Inventory is a required field');
    }
  };

  const validateFields = (name: string, inventory: string) => {
    try {
      validateName(name);
      validateInventory(inventory);
    } catch (err: any) {
      throw new Error(err);
    }
  };

  const handleSubmit = async (name: string, inventory: string, description?: string | undefined) => {
    try {
      setIsSubmitting(true);
      validateFields(name, inventory);
      addInventory({ name, description, inventory }).then((data) => {
        data?.data?.id ? history.push(`/inventories/${data?.data?.id}/details`) : history.push(`/inventories`);
        dispatch(
          addNotification({
            variant: 'success',
            title: intl.formatMessage(sharedMessages.addInventory),
            dismissable: true,
            description: intl.formatMessage(sharedMessages.add_inventory_success),
          })
        );
      });
    } catch (error) {
      dispatch(
        addNotification({
          variant: 'danger',
          title: intl.formatMessage(sharedMessages.addInventory),
          dismissable: true,
          description: `${intl.formatMessage(sharedMessages.add_inventory_failure)}  ${error}`,
        })
      );
    }
  };

  return (
    <React.Fragment>
      <TopToolbar
        breadcrumbs={[
          {
            title: 'Inventories',
            to: '/inventories',
            key: 'inventories',
          },
          {
            title: 'Add',
            key: 'inventory-add',
          },
        ]}
      >
        <Title headingLevel={'h2'}>Add new inventory</Title>
      </TopToolbar>
      <PageSection>
        <Card>
          <CardBody>
            <Form>
              <InventoryForm
                handleSubmit={(name, inventory, description) => handleSubmit(name, inventory, description)}
                isSubmitting={isSubmitting}
              />
            </Form>
          </CardBody>
        </Card>
      </PageSection>
    </React.Fragment>
  );
};

export { NewInventory };
