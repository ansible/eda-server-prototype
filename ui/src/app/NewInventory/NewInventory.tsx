import {
  ActionGroup,
  Button,
  Card,
  CardBody as PFCardBody,
  Form,
  FormGroup,
  Grid,
  GridItem,
  PageSection,
  Stack,
  StackItem,
  TextInput,
  Title,
  ValidatedOptions,
} from '@patternfly/react-core';
import { useHistory } from 'react-router-dom';
import React, { useState } from 'react';
import styled from 'styled-components';
import { TopToolbar } from '@app/shared/top-toolbar';
import { ExclamationCircleIcon } from '@patternfly/react-icons';
import { useIntl } from 'react-intl';
import sharedMessages from '../messages/shared.messages';
import { addNotification } from '@redhat-cloud-services/frontend-components-notifications';
import { useDispatch } from 'react-redux';
import { addInventory } from '@app/API/Inventory';
import { EdaCodeEditor } from '@app/utils/EdaCodeEditor';
import { Language } from '@patternfly/react-code-editor';

const CardBody = styled(PFCardBody)`
  white-space: pre-wrap;
`;
const NewInventory: React.FunctionComponent = () => {
  const history = useHistory();
  const intl = useIntl();
  const dispatch = useDispatch();
  const [name, setName] = useState('');
  const [description, setDescription] = useState('');
  const [inventory, setInventory] = useState('');
  const [validatedName, setValidatedName] = useState<ValidatedOptions>(ValidatedOptions.default);
  const [validatedInventory, setValidatedInventory] = useState<ValidatedOptions>(ValidatedOptions.default);
  const [isSubmitting, setIsSubmitting] = useState<boolean>(false);

  const validateName = (value) => {
    !value || value.length < 1 ? setValidatedName(ValidatedOptions.error) : setValidatedName(ValidatedOptions.default);
  };

  const onNameChange = (value) => {
    setName(value);
    validateName(value);
  };

  const onDescriptionChange = (value) => {
    setDescription(value);
  };

  const validateInventory = (value) => {
    !value || value.length < 1
      ? setValidatedInventory(ValidatedOptions.error)
      : setValidatedInventory(ValidatedOptions.default);
  };

  const onInventoryChange = (value) => {
    setInventory(value);
    validateInventory(value);
  };

  const validateFields = () => {
    validateName(name);
    validateInventory(inventory);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    validateFields();
    addInventory({ name: name, description: description, inventory: code })
      .then((data) => {
        data?.data?.id ? history.push(`/inventories/inventory/${data?.data?.id}`) : history.push(`/inventories`);
        dispatch(
          addNotification({
            variant: 'success',
            title: intl.formatMessage(sharedMessages.addInventory),
            dismissable: true,
            description: intl.formatMessage(sharedMessages.add_inventory_success),
          })
        );
      })
      .catch((error) => {
        history.push(`/inventories`);
        dispatch(
          addNotification({
            variant: 'danger',
            title: intl.formatMessage(sharedMessages.addInventory),
            dismissable: true,
            description: `${intl.formatMessage(sharedMessages.add_inventory_failure)}  ${error}`,
          })
        );
      });
  };

  const onEditorDidMount = (editor, monaco) => {
    // eslint-disable-next-line no-console
    editor.focus();
  };

  const [code, setCode] = React.useState('');

  return (
    <React.Fragment>
      <TopToolbar
        breadcrumbs={[
          {
            title: 'Inventories',
            to: '/inventories',
          },
          {
            title: 'Add',
          },
        ]}
      >
        <Title headingLevel={'h2'}>Add new inventory</Title>
      </TopToolbar>
      <PageSection>
        <Card>
          <CardBody>
            <Form>
              <Stack hasGutter style={{ paddingRight: '30px' }}>
                <StackItem>
                  <Grid hasGutter>
                    <GridItem span={4}>
                      <FormGroup
                        style={{ paddingRight: '30px' }}
                        label={intl.formatMessage(sharedMessages.name)}
                        fieldId={`inventory-name`}
                        isRequired
                        helperTextInvalid={intl.formatMessage(sharedMessages.enterInventoryName)}
                        helperTextInvalidIcon={<ExclamationCircleIcon />}
                        validated={validatedName}
                      >
                        <TextInput
                          id="inventory-name"
                          value={name}
                          label="Name"
                          isRequired
                          validated={validatedName}
                          onChange={onNameChange}
                          onBlur={(event) => validateName(name)}
                          placeholder={intl.formatMessage(sharedMessages.namePlaceholder)}
                        />
                      </FormGroup>
                    </GridItem>
                    <GridItem span={4}>
                      <FormGroup
                        style={{ paddingRight: '30px' }}
                        label={intl.formatMessage(sharedMessages.description)}
                        fieldId={`inventory-description`}
                        helperTextInvalid={intl.formatMessage(sharedMessages.enterInventoryDescription)}
                        helperTextInvalidIcon={<ExclamationCircleIcon />}
                      >
                        <TextInput
                          id="inventory-description"
                          label="Description"
                          placeholder={intl.formatMessage(sharedMessages.descriptionPlaceholder)}
                          onChange={onDescriptionChange}
                        />
                      </FormGroup>
                    </GridItem>
                  </Grid>
                </StackItem>
                <StackItem>
                  <FormGroup
                    label={intl.formatMessage(sharedMessages.inventory)}
                    fieldId={`inventory-inventory`}
                    width={'80pct'}
                  >
                    <EdaCodeEditor
                      code={code}
                      editMode={true}
                      setEditedCode={setCode}
                      language={Language.yaml}
                      width={'100%'}
                      height={'400'}
                    />
                  </FormGroup>
                </StackItem>
                <StackItem>
                  <ActionGroup>
                    <Button variant="primary" onClick={handleSubmit} isLoading={isSubmitting}>
                      {isSubmitting ? 'Adding ' : 'Add'}
                    </Button>
                    <Button variant="link" onClick={() => history.push('/inventories')}>
                      {' '}
                      Cancel
                    </Button>
                  </ActionGroup>
                </StackItem>
              </Stack>
            </Form>
          </CardBody>
        </Card>
      </PageSection>
    </React.Fragment>
  );
};

export { NewInventory };
