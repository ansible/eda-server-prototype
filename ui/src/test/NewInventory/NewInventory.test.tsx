import * as React from 'react';
import { mount, render, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router';
import { act } from 'react-dom/test-utils';
import { IntlProvider } from 'react-intl';
import { Route } from 'react-router-dom';
import { Button, TextInput } from '@patternfly/react-core';
import store from '../../store';
import { Provider } from 'react-redux';
import { NewInventory } from '@app/NewInventory/NewInventory';
import { mockApi } from '../__mocks__/baseApi';

const ComponentWrapper = ({ children, initialEntries = ['/dashboard'] }) => (
  <Provider store={store()}>
    <IntlProvider locale="en">
      <MemoryRouter initialEntries={initialEntries} keyLength={0} key={'test'}>
        {children}
      </MemoryRouter>
    </IntlProvider>
  </Provider>
);

describe('NewInventory', () => {
  beforeAll(() => {
    mockApi.reset();
    window.HTMLCanvasElement.prototype.getContext = () => ({} as any);
  });

  it('should render the NewInventory form', async () => {
    mockApi.onPost(`/api/inventories`).replyOnce(200, {
      name: 'Inventory 1',
      id: 1,
      inventory: 'inventory',
    });

    let wrapper;
    await act(async () => {
      wrapper = render(
        <ComponentWrapper initialEntries={['/new-inventory']}>
          <Route path="/new-inventory">
            <NewInventory />
          </Route>
        </ComponentWrapper>
      );
    });

    expect(wrapper).toMatchSnapshot();
  });
});
