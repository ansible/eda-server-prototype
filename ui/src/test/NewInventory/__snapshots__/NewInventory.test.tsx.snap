// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewInventory should render the NewInventory form 1`] = `
[
  <section
    class="pf-c-page__main-section pf-m-light"
  >
    <div
      class="pf-l-level"
    >
      <div>
        <nav
          aria-label="Breadcrumb"
          class="pf-c-breadcrumb"
          data-ouia-component-id="OUIA-Generated-Breadcrumb-1"
          data-ouia-component-type="PF4/Breadcrumb"
          data-ouia-safe="true"
        >
          <ol
            class="pf-c-breadcrumb__list"
          >
            <li
              class="pf-c-breadcrumb__item"
              title="Inventories"
            >
              <a
                href="/inventories"
              >
                Inventories
              </a>
            </li>
            <li
              class="pf-c-breadcrumb__item"
              title="Add"
            >
              <span
                class="pf-c-breadcrumb__item-divider"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align:-0.125em"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </span>
              Add
            </li>
          </ol>
        </nav>
      </div>
    </div>
    <h2
      class="pf-c-title pf-m-xl"
      data-ouia-component-id="OUIA-Generated-Title-1"
      data-ouia-component-type="PF4/Title"
      data-ouia-safe="true"
    >
      Add new inventory
    </h2>
  </section>,
  <section
    class="pf-c-page__main-section"
  >
    <article
      class="pf-c-card"
      data-ouia-component-id="OUIA-Generated-Card-1"
      data-ouia-component-type="PF4/Card"
      data-ouia-safe="true"
      id=""
    >
      <div
        class="pf-c-card__body sc-bcXHqe FaSym"
      >
        <form
          class="pf-c-form"
          novalidate=""
        >
          <div
            class="pf-l-stack pf-m-gutter"
            style="padding-right:30px"
          >
            <div
              class="pf-l-stack__item"
            >
              <div
                class="pf-l-grid pf-m-gutter"
              >
                <div
                  class="pf-l-grid__item pf-m-4-col"
                >
                  <div
                    class="pf-c-form__group"
                    style="padding-right:30px"
                  >
                    <div
                      class="pf-c-form__group-label"
                    >
                      <label
                        class="pf-c-form__label"
                        for="inventory-name"
                      >
                        <span
                          class="pf-c-form__label-text"
                        >
                          Name
                        </span>
                        <span
                          aria-hidden="true"
                          class="pf-c-form__label-required"
                        >
                           *
                        </span>
                      </label>
                       
                    </div>
                    <div
                      class="pf-c-form__group-control"
                    >
                      <input
                        aria-invalid="false"
                        class="pf-c-form-control"
                        data-ouia-component-id="OUIA-Generated-TextInputBase-1"
                        data-ouia-component-type="PF4/TextInput"
                        data-ouia-safe="true"
                        id="inventory-name"
                        label="Name"
                        placeholder="Insert name here"
                        required=""
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="pf-l-grid__item pf-m-4-col"
                >
                  <div
                    class="pf-c-form__group"
                    style="padding-right:30px"
                  >
                    <div
                      class="pf-c-form__group-label"
                    >
                      <label
                        class="pf-c-form__label"
                        for="inventory-description"
                      >
                        <span
                          class="pf-c-form__label-text"
                        >
                          Description
                        </span>
                      </label>
                       
                    </div>
                    <div
                      class="pf-c-form__group-control"
                    >
                      <input
                        aria-invalid="false"
                        class="pf-c-form-control"
                        data-ouia-component-id="OUIA-Generated-TextInputBase-2"
                        data-ouia-component-type="PF4/TextInput"
                        data-ouia-safe="true"
                        id="inventory-description"
                        label="Description"
                        placeholder="Insert description here"
                        type="text"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="pf-l-stack__item"
            >
              <div
                class="pf-c-form__group"
                width="80pct"
              >
                <div
                  class="pf-c-form__group-label"
                >
                  <label
                    class="pf-c-form__label"
                    for="inventory-inventory"
                  >
                    <span
                      class="pf-c-form__label-text"
                    >
                      Inventory
                    </span>
                  </label>
                   
                </div>
                <div
                  class="pf-c-form__group-control"
                >
                  <div
                    class="pf-c-code-editor"
                  >
                    <div
                      class="pf-c-code-editor__header"
                    >
                      <div
                        class="pf-c-code-editor__controls"
                      />
                      <div
                        class="pf-c-code-editor__header-main"
                      />
                    </div>
                    <div
                      class="pf-c-code-editor__main"
                    >
                      <div
                        class="pf-c-code-editor__code"
                        tabindex="0"
                      >
                        <div
                          class="react-monaco-editor-container"
                          style="width:100%;height:400px"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="pf-l-stack__item"
            >
              <div
                class="pf-c-form__group pf-m-action"
              >
                <div
                  class="pf-c-form__group-control"
                >
                  <div
                    class="pf-c-form__actions"
                  >
                    <button
                      aria-disabled="false"
                      class="pf-c-button pf-m-primary pf-m-progress"
                      data-ouia-component-id="OUIA-Generated-Button-primary-1"
                      data-ouia-component-type="PF4/Button"
                      data-ouia-safe="true"
                      type="button"
                    >
                      Add
                    </button>
                    <button
                      aria-disabled="false"
                      class="pf-c-button pf-m-link"
                      data-ouia-component-id="OUIA-Generated-Button-link-1"
                      data-ouia-component-type="PF4/Button"
                      data-ouia-safe="true"
                      type="button"
                    >
                       Cancel
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
    </article>
  </section>,
]
`;
